src/docker.ts

```diff
@@ -62,8 +62,12 @@ export async function loginECR(registry: string, username: string, password: str
     core.info(`AWS ECR detected with ${region} region`);
   }
 
-  process.env.AWS_ACCESS_KEY_ID = username || process.env.AWS_ACCESS_KEY_ID;
-  process.env.AWS_SECRET_ACCESS_KEY = password || process.env.AWS_SECRET_ACCESS_KEY;
+  if (username) {
+    process.env.AWS_ACCESS_KEY_ID = username;
+  }
+  if (password) {
+    process.env.AWS_SECRET_ACCESS_KEY = password;
+  }
```

Comment:
This logic is the same thing as the current one: https://onecompiler.com/nodejs/3xmb4n6tq

Reply:
The outcome is sadly different

```js
const util = require('util');
const exec = util.promisify(require('child_process').exec);
async function printenv() {
    process.env.TEST_ENV_BANANA = undefined
    process.env.TEST_ENV_APPLE = undefined
    delete process.env.TEST_ENV_APPLE
  try {
      const { stdout, stderr } = await exec('printenv | grep TEST_ENV_');
      console.log('stdout:', stdout);
      console.log('stderr:', stderr);
  } catch (err) {
     console.error(err);
  };
};
printenv();
```

![image](https://user-images.githubusercontent.com/326929/146008355-f41ce1e9-d698-4af1-aa10-de67fc7cdf86.png)

As you can see in the case of `TEST_ENV_BANANA` the environment variable is set in the subprocess.
This causes issue when using EC2 metadata api for getting AWS credentials, as now the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` will both be set to `undefined` eg they have a value and therefor the AWS cli will try to use them instead of falling back thru the credential chain and use the ec2 metadata api.

Reply:
So the problem is when 
```js
 process.env.AWS_SECRET_ACCESS_KEY = password ||  process.env.AWS_SECRET_ACCESS_KEY
 ```
 
 if `process.env.AWS_SECRET_ACCESS_KEY` and `password` are both `undefined` we end up assigning `process.env.AWS_SECRET_ACCESS_KEY = undefined` which causes this env var to be set for subprocesses.

Reply:
Here's an updated onecompiler link where you can see that the result will differ
https://onecompiler.com/nodejs/3xmbfxhf6
https://onecompiler.com/nodejs/3xmbg2nzw

Reply:
Oh ok so this is related to sub steps in the workflow right?

Reply:
Yes!
If you want to use this action on a self-hosted runner on AWS, the runner can have access to login to AWS ECR without having to set any credentials

```yaml
      - name: Login to ECR
        uses: docker/login-action@v1
        with:
          registry: <aws-account-number>.dkr.ecr.<region>.amazonaws.com
```

Eventually the action will run the aws cli to get the login token, line 77 or 81 here. Which would be affected by any `process.env` changes done in node here since it a child process
https://github.com/docker/login-action/blob/1cce1654e01caa87cc0eac1c100e6749708affb9/src/aws.ts#L74-L85

We use the docker/login-action for or other workflows so it would be nice to be able to use it on our self-hosted runners as well without having to specify the credentials in those workflows :)

Why node.js propagates env vars as undefined I have no clue 😄

Reply:
Ok I think instead we could set the env var to the current process in `getExecOutput`. I will open a PR with this solution and come back to you. I keep yours open in case mine doesn't fit your needs.